accepted/2021/runtime-security-mitigations.md

```diff
@@ -0,0 +1,213 @@
+
+## Stack Buffer Overflow Protection
+
+[Stack Buffer Overflow](https://en.wikipedia.org/wiki/Stack_buffer_overflow) protection is among the older security mitigations. It has been known as &quot;GS cookie&quot; on Windows. It is designed to catch writes into stack memory outside the intended bounds.
+
+The first .NET runtime that included it was .NET Framework 2.0, and the implementation has not changed much since then. The mitigation in C/C++ compilers have seen improvements in the meantime (for example, [Enhanced GS](https://msrc-blog.microsoft.com/2009/03/20/enhanced-gs-in-visual-studio-2010/) or -[fstack-protector-strong](https://lwn.net/Articles/584225/)).
+
+We will review and update stack buffer overflow protection in .NET runtime to match state of the art.
+
+# Language, Libraries and Runtime Features
+
+Library and language design have always played a key role in mitigating security bugs by preventing or discouraging them from being introduced in the first place.
+
+## Unsafe API and C# language constructs
+
+Number of newer APIs such as `Memory<T>` and friends come with security risks that are not clearly communicated.
+
+We need to adopt a more holistic approach to unsafe features and APIs that we expose for public use. Developers need feedback in the IDE so that they understand that they are using constructs that are unsafe and are [informed of the specific risks that come with using those APIs](https://github.com/dotnet/runtime/issues/41418). Over time, we bias more to constructs and paradigms that are safe by default (possibly inspired by Rust).
+
+## Object and Array Pooling
+
+`ArrayPool<T>` and other similar custom object pools are equivalent of the classic C malloc. Manual lifetime management traditionally suffers from use-after-free and double-free security vulnerabilities. Many years were spent on implementing diagnostics and mitigations of these vulnerabilities for malloc.
+
+We need to implement similar diagnostic and mitigation capabilities for ArrayPool\&lt;T\&gt; and friends. It is [not unusual](https://github.com/dotnet/corefx/pull/37270) to find security bugs caused by incorrect `ArrayPool<T>` use even in dotnet/runtime libraries.
+
+We are also going to be continually pushed to expose APIs like those we use internally which make potential misuse even easier, e.g. `ValueStringBuilder`. Thus far we have simply said &quot;we're not comfortable doing that&quot;, but it is unlikely going to be a satisfactory answer long-term. It is something we need to keep thinking about.
```

Comment:
It's a satisfactory answer for me :) More seriously satisfactory is a value judgement. How about 

_We are continually pushed to expose potentially dangerous APIs we use internally, e.g. `ValueStringBuilder`. Thus far we have simply said &quot;we're not comfortable doing that&quot;, and we will continue to examine the balance of security against utility". If an API can be safe by default it is more likely to be included._

Reply:
With `ValueStringBuilder` specifically, I think we said we're comfortable shipping it once we get good language guardrails for its usage. C# doesn't have a strong ownership model, but we could reconsider that. We could also introduce copy-only structs (see https://gist.github.com/GrabYourPitchforks/7ab6a440100467a82cfe5998cd1e91be for a basic proposal), or we could write a Roslyn analyzer to govern `ValueStringBuilder` usage.

So I wouldn't say we're _opposed_ to doing it, just that we're waiting for some changes in the ecosystem before we introduce it publicly.

Reply:
Even if it's not secure by default, if we expose an API that's potentially dangerous, we should aim to simultaneously offer an analyzer that makes it significantly harder to use insecurely.

Reply:
I expect in .NET 6 we're going to ship something like it in support of interpolated strings, but it'll be buried in System.Runtime.CompilerServices; while someone could use it directly, it'll fit in with the rest of System.Runtime.CompilerServices and System.Runtime.InteropServices as "if you touch this stuff, it's unsafe and you need to know what you're doing".

Reply:
>  move-only structs

Do we have one central issue to point to that describes the problem and discusses potential designs?

Reply:
Opened https://github.com/dotnet/runtime/issues/50389